example/src/app/index.tsx

```diff
@@ -1,11 +1,40 @@
 import * as React from 'react';
 import * as ReactDOM from 'react-dom';
-
-import { Provider, createClient } from '../../../src/index';
+import {
+  Provider,
+  createClient,
+  dedupeExchange,
+  fetchExchange,
+  cacheExchange,
+  createSubscriptionExchange,
+} from '../../../src/index';
 import Home from './home';
 
+import { SubscriptionClient } from 'subscriptions-transport-ws';
+const subscriptionClient = new SubscriptionClient(
```

Comment:
Would creating the websocket client inside the subscriptionExchange work? Ideally the complexity of the websocket communication would be abstracted.

I'm thinking this way we can remove the need for users to have to configure their websocket connection and subscription exchange. In the case where no subscriptions are created within the app, no websocket connection would be made.

```tsx
// ..
import { WebSocket } from 'ws'; // Or whatever websocket library we use

// ..
export const subscriptionExchange: Exchange = ({ forward }) => {
  /** Store of subscription operations */
  let operations = new Map<string, Operation>();
  /** Store of response subjects */
  let subjects = new Map<string, Subject<ExchangeResult>>();
  /** Websocket connection */
  let socket: WebSocket | undefined;

  /** Create connection on first subscription operation if not found */
  const getConnection = (target: string) => {
    if (socket !== undefined) {
      return socket;
    }

    socket = new WebSocket(target);

    socket.on('message', (data: any) => {
      const subscriptionName = getSubscriptionNameFromData(data); // Somehow map response to subscription id?
      const sub = subjects.get(subscriptionName);

      sub.next({ data, operation: operations.get(subscriptionName)}); // Send response back to client instance
    });

    return socket;
  }

  const useSubscription = (operation: Operation) => {
    const connectedSocket = getConnection(operation.context.websocketUrl);
    const subscriptionName = getSubscriptionNameFromOperation(operation); // Somehow parse subscription for unique id

    if (!operations.has(subscriptionName)) {
      operations.set(subscriptionName, operation);
      subjects.set(subscriptionName, new Subject<ExchangeResult>());
    }

    connectedSocket.send(operation.query); // Send query over socket
    return subjects.get(subscriptionName).asObservable(); // Return stream (new data will be pushed by socket.on('message'))
  }

  return ops$.pipe(
    mergeMap(operation => operation.operationName !== 'subscription' ? forward(of(operation)) : useSubscription(operation))
  )
}
```

Reply:
Sorry for leaving a third rather negative comment 😅

The HTTP spec part of GraphQL is so narrow (unintentionally) that there's not really an alternative to it. Thus all http / fetch logic across all graphql client libraries end up looking roughly the same. The same cannot be said about websockets.

Since the GraphQL spec makes no assumptions about transport layers there's no standard to graphql over websockets. This means that the transport implementation is specific and coupled to how the server implents it. The Apollo libraries have corresponding libs for the server side and client side for subscriptions.

So the motivation behind the pre-v1 api was to keep this agnostic so any transport layer implementation can be used for subscriptions. It wouldn't make sense to couple this unless we assume a specific implementation on the server-side, which is probably a job for another library.

Reply:
Wait... So Graphql subscription/websocket implementations are not standardized (yet)?

I'd definitely still prefer if we can keep exchanges plug an play rather than having to create them.

Maybe we could pass send and receive callbacks to the client and then forward them through to the exchange - that way we only 'instantiate'/configure via the client.

Reply:
Couple thoughts:

1. yes, they are non-standard, so @kitten is right in his synopsis.
2. if we were to try and bundle as much in this library as possible we could potentially bloat the lib for non-subscription users (majority?).
3. Isn't the point of exchanges to be configurable? I think putting the subscription exchange in the defaultExchanges might be unwise since it's not as common a use case and increases bytes.

Reply:
I think exchanges are useful but being able to abstract the concept for core functionality, including subscriptions, would be useful to most. 

The point of exchanges (as I see it) is for extensibility - to take our implementation of _urql_ and tweak/add to it as you please. Something more vanilla such as subscriptions probably doesn't need the involvement of our extensibility interface?

How about the following - I think it addresses the concerns mentioned but while keeping things simple and consistent:

 * Default exchanges are: _subscription, dedupe, cache, fetch_.
 * Users configure default exchanges at _client_ creation (as currently is the case with fetch config).
 * Subscription config:
   * Users configure their own communication client (addresses non-standard problem prevents aforementioned additional bloat) and provide utility functions to the _urql_ client for sending and receiving websocket communication.
   * The _subscription exchange_ throws a runtime error (or console error message in development) if a subscription operation is called without the required config options being supplied to the _client_.

Reply:
Frankly I'll defer to you and @kitten. I think there are valid arguments on both sides and I don't feel like the right person to make that decision since i'm less involved with the codebase as a whole.

If you guys can agree on something I'll make whatever change that is.

Reply:
Cool. I'm changing this then to make a subscriptionExchange part of the defaultExchanges, and adding an API option to client init called `forwardOperation`